components/tree/index.jsx

```diff
@@ -27,10 +27,38 @@ import { TREE } from '../../utilities/constants';
  * A tree is visualization of a structure hierarchy. A branch can be expanded or collapsed. This is a controlled component, since visual state is present in the `nodes` data.
  */
 class Tree extends React.Component {
+	constructor (props) {
+		super(props);
+		this.state = {
+			flattenedNodes: this.flattenTree({ nodes: this.props.nodes, expanded: true }).slice(1)
+		};
+	}
+
 	componentWillMount () {
 		checkProps(TREE, this.props);
 	}
 
+	componentWillReceiveProps (nextProps) {
+		this.state = {
+			flattenedNodes: this.flattenTree({ nodes: nextProps.nodes, expanded: true }).slice(1)
+		};
+	}
+
+	// Flattens hierarchical tree structure into a flat array.
+	flattenTree (root, treeIndex = '') {
```

Comment:
This seems to be performant enough with small trees. There's a large dataset story, but it's already some slowness with select/unselect, so I think it's the DOM update and not any kind of data within the component.